### PR TITLE
Add in access list member backend and gRPC methods.

### PR DESCRIPTION
**Pull request by mdwn**
The backend and gRPC methods for interacting with access lists have been added. This additionally requires modifications to the generic service to handle key prefixes so that access list members can be namespaced properly within the access_list_member namespace in the backend.

### CODE REVIEW

File: lib/services/access_list.go [hunk 1]
```diff
@@ -96,6 +96,30 @@ func UnmarshalAccessList(data []byte, opts ...MarshalOption) (*accesslist.Access
 	return accessList, nil
 }
 
+// AccessListMembersGetter defines an interface for reading access list members.
+type AccessListMembersGetter interface {
+	// GetAccessListMembers returns the members associated with the access list resource.
+	GetAccessListMembers(context.Context, string) ([]*accesslist.AccessListMember, error)
+	// ListAccessListMembers returns a paginated list of all access list members.
+	ListAccessListMembers(context.Context, string, int, string) ([]*accesslist.AccessListMember, string, error)
```
Comment (codingllama) [inline lib/services/access_list.go:104]: `string, int, string` is a pretty ambiguous sequence of parameters.

File: lib/services/local/access_list.go [hunk 2]
```diff
@@ -31,21 +31,24 @@ import (
...
+	log    logrus.FieldLogger
+	clock  clockwork.Clock
+	alSvc  *generic.Service[*accesslist.AccessList]
+	almSvc *generic.Service[*accesslist.AccessListMember]
 }
 
 // NewAccessListService creates a new AccessListService.
 func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*AccessListService, error) {
-	svc, err := generic.NewService(&generic.ServiceConfig[*accesslist.AccessList]{
+	alSvc, err := generic.NewService(&generic.ServiceConfig[*accesslist.AccessList]{
```
Comment (codingllama) [inline lib/services/local/access_list.go:51]:
```suggestion
	service, err := generic.NewService(&generic.ServiceConfig[*accesslist.AccessList]{
```

File: lib/services/local/access_list.go [hunk 3]
```diff
@@ -57,44 +60,109 @@ func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*Acce
 		return nil, trace.Wrap(err)
 	}
 
+	almSvc, err := generic.NewService(&generic.ServiceConfig[*accesslist.AccessListMember]{
```
Comment (codingllama) [inline lib/services/local/access_list.go:63]:
```suggestion
	memberService, err := generic.NewService(&generic.ServiceConfig[*accesslist.AccessListMember]{
```

File: lib/services/local/access_list.go [hunk 4]
```diff
@@ -31,21 +31,24 @@ import (
...
 
 // AccessListService manages Access List resources in the Backend.
 type AccessListService struct {
-	log   logrus.FieldLogger
-	clock clockwork.Clock
-	svc   *generic.Service[*accesslist.AccessList]
+	log    logrus.FieldLogger
+	clock  clockwork.Clock
+	alSvc  *generic.Service[*accesslist.AccessList]
+	almSvc *generic.Service[*accesslist.AccessListMember]
```
Comment (codingllama) [inline lib/services/local/access_list.go:46]: Ditch the unusual abbreviations? The larger the scope, the clearer the name should be.

```suggestion
	service  *generic.Service[*accesslist.AccessList]
	memberService *generic.Service[*accesslist.AccessListMember]
```

File: lib/services/local/access_list.go [hunk 5]
```diff
@@ -57,44 +60,109 @@ func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*Acce
...
 // DeleteAccessList removes the specified access list resource.
 func (a *AccessListService) DeleteAccessList(ctx context.Context, name string) error {
-	return trace.Wrap(a.svc.DeleteResource(ctx, name))
+	err := a.alSvc.DeleteResource(ctx, name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Delete all associated members as well.
+	return trace.Wrap(a.DeleteAllAccessListMembersForAccessList(ctx, name))
```
Comment (codingllama) [inline lib/services/local/access_list.go:116]: What are the consequences of succeeding the first request and failing this one? Is that something we should be concerned about?

File: lib/services/local/access_list.go [hunk 6]
```diff
@@ -57,44 +60,109 @@ func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*Acce
...
 // DeleteAllAccessLists removes all access lists.
 func (a *AccessListService) DeleteAllAccessLists(ctx context.Context) error {
-	return trace.Wrap(a.svc.DeleteAllResources(ctx))
+	err := trace.Wrap(a.alSvc.DeleteAllResources(ctx))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Delete all members for all access lists.
+	return trace.Wrap(a.DeleteAllAccessListMembers(ctx))
```
Comment (codingllama) [inline lib/services/local/access_list.go:127]: Same as above?

File: lib/services/local/access_list.go [hunk 7]
```diff
@@ -57,44 +60,109 @@ func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*Acce
...
+	return trace.Wrap(a.DeleteAllAccessListMembersForAccessList(ctx, name))
 }
 
 // DeleteAllAccessLists removes all access lists.
 func (a *AccessListService) DeleteAllAccessLists(ctx context.Context) error {
-	return trace.Wrap(a.svc.DeleteAllResources(ctx))
+	err := trace.Wrap(a.alSvc.DeleteAllResources(ctx))
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
Comment (codingllama) [inline lib/services/local/access_list.go:124]: Why double-wrapping?

```suggestion
	if err := a.alSvc.DeleteAllResources(ctx); err != nil {
		return trace.Wrap(err)
	}
```

File: lib/services/local/access_list.go [hunk 8]
```diff
@@ -57,44 +60,109 @@ func NewAccessListService(backend backend.Backend, clock clockwork.Clock) (*Acce
...
+	member, err := a.almSvc.GetResource(ctx, memberName, accessList)
+	return member, trace.Wrap(err)
+}
+
+// UpsertAccessListMember creates or updates an access list member resource.
+func (a *AccessListService) UpsertAccessListMember(ctx context.Context, member *accesslist.AccessListMember) (*accesslist.AccessListMember, error) {
+	if err := trace.Wrap(a.almSvc.UpsertResource(ctx, member, member.Spec.AccessList)); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return member, nil
```
Comment (codingllama) [inline lib/services/local/access_list.go:152]:
```suggestion
	if err := a.almSvc.UpsertResource(ctx, member, member.Spec.AccessList); err != nil {
		return nil, trace.Wrap(err)
	}
	return member, nil
```

Comment (codingllama) [inline lib/services/local/access_list.go:152]: As a general design comment, it looks off to me that service.UpsertResource doesn't return the member, but this one does. I'd rather the service did it too.

File: lib/services/local/access_list_test.go [hunk 9]
```diff
@@ -106,13 +106,113 @@ func TestAccessListCRUD(t *testing.T) {
 	require.NoError(t, err)
 	require.Empty(t, cmp.Diff(accessList1, accessList, cmpOpts...))
 
+	// There should be no access list members
+	members, err := service.GetAccessListMembers(ctx, accessList1.GetName())
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:110]: I think this test is definitely doing too much now.

File: lib/services/local/generic/generic.go [hunk 10]
```diff
@@ -280,8 +282,15 @@ func (s *Service[T]) MakeBackendItem(resource T, name string) (backend.Item, err
...
-	return backend.Key(s.backendPrefix, name)
+func (s *Service[T]) MakeKey(name string, parts ...string) []byte {
+	numKeyParts := len(parts)
+	keyParts := make([]string, numKeyParts+2)
+	keyParts[0] = s.backendPrefix
+	keyParts[numKeyParts+1] = name
+	if numKeyParts > 0 {
+		copy(keyParts[1:], parts)
+	}
+	return backend.Key(keyParts...)
```
Comment (codingllama) [inline lib/services/local/generic/generic.go:293]:
```suggestion
	return backend.Key(append([]string{s.backendPrefix, name}, parts...)...)
```

(There may be a more elegant way.)

File: lib/services/local/generic/generic_test.go [hunk 11]
```diff
@@ -264,3 +264,49 @@ func TestGenericCRUD(t *testing.T) {
...
+				Backend:       memBackend,
+				ResourceKind:  "generic resource",
+				PageLimit:     200,
+				BackendPrefix: backendPrefix,
+				UnmarshalFunc: unmarshalResource,
+				MarshalFunc:   marshalResource,
+			})
+			require.NoError(t, err)
+
+			key := service.MakeKey(test.keyName, test.prefix...)
```
Comment (codingllama) [inline lib/services/local/generic/generic_test.go:308]: Suggestion: Implement `service.MakeKey` using a top-level `func MakeKey(prefix, name string, parts ...string)`, so we can test it without all the service setup?

If one just calls the other then visual inspection is enough. (Plus it gets exercised a bunch.)

That is, assuming we even want to keep the test?

**Review by codingllama — COMMENTED** [review]
My 2c: I prefer to split protos and implementation changes into separate PRs, it keeps discussion focused on just design or impl.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 12]
```diff
@@ -29,22 +29,28 @@ service AccessListService {
...
-  // UpsertAccessListMember creates or updates an access list member resource.
-  rpc UpsertAccessListMember(UpsertAccessListMemberRequest) returns (Member);
   // DeleteAccessList hard deletes the specified access list resource.
   rpc DeleteAccessList(DeleteAccessListRequest) returns (google.protobuf.Empty);
   // DeleteAccessListMember hard deletes the specified access list member resource.
   rpc DeleteAccessListMember(DeleteAccessListMemberRequest) returns (google.protobuf.Empty);
   // DeleteAllAccessLists hard deletes all access lists.
   rpc DeleteAllAccessLists(DeleteAllAccessListsRequest) returns (google.protobuf.Empty);
+
+  // GetAccessListMembers returns the members associated with the access list resource.
```
[the remark this answers is not in the document]
Comment (mdwn) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:41]: I'll remove it here, having a paginated list is fine.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 13]
```diff
@@ -77,6 +83,23 @@ message GetAccessListRequest {
...
+// UpsertAccessListRequest is the request for upserting an access list.
+message UpsertAccessListRequest {
+  // access_list is the access list to upsert.
+  AccessList access_list = 1;
+}
+
+// DeleteAccessListRequest is the request for deleting an access list.
+message DeleteAccessListRequest {
+  // name is the name of the access list to delete.
+  string name = 1;
```
[the remark this answers is not in the document]
Comment (mdwn) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:95]: They are and they are always identified b y name.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 14]
```diff
@@ -89,10 +112,30 @@ message GetAccessListMembersResponse {
...
+  string next_token = 2;
+}
+
+// GetAccessListMemberRequest is the request for retrieving an access list member.
+message GetAccessListMemberRequest {
+  // access_list is the name of the access list that the member belongs to.
+  string access_list = 1;
+
+  // name is the name of the user that belongs to the access list.
+  string name = 2;
```
[the remark this answers is not in the document]
Comment (mdwn) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:138]: The new messages are okay to modify, I think, but the older AccessList methods can't be. I'll make the changes.

File: lib/services/local/access_list.go (hunk 5 above)
Comment (mdwn) [reply lib/services/local/access_list.go:116]: There's a potential for there to be leaked access list member objects. It's not ideal, but I'm not 100% sure the way around this without transactions.

There is a potential for users to be inadvertently added to a new access list if it's named the same as the old access list, e.g.

* Access list is deleted, member delete fails.
* Access list is recreated, members already map to new list.

We can get around this by doing the delete the other way around, I think. I'll make that change. Same applies to below.

File: lib/services/local/access_list.go (hunk 7 above)
Comment (mdwn) [reply lib/services/local/access_list.go:124]: Bad copy-paste! Fixed.

File: lib/services/local/generic/generic.go (hunk 10 above)
Comment (mdwn) [reply lib/services/local/generic/generic.go:293]: This has the potential to return unexpected objects when doing pagination -- I think the extra namespacing bits need to be prefixed prior to the name here.

File: lib/services/local/generic/generic_test.go (hunk 11 above)
Comment (mdwn) [reply lib/services/local/generic/generic_test.go:308]: Seems reasonable to me.

Comment (mdwn) [reply lib/services/local/generic/generic_test.go:308]: Oh actually, I remember now -- this is on purpose so that callers can use MakeKey and be assured that the associated backendPrefix is pushed into place. I'd rather leave that as is for that reason.

File: lib/services/local/access_list_test.go (hunk 9 above)
Comment (mdwn) [reply lib/services/local/access_list_test.go:110]: Working on trying to split this out.

File: lib/services/local/generic/generic.go (hunk 10 above)
Comment (codingllama) [reply lib/services/local/generic/generic.go:293]: Could we then change the signature to `MakeKey(names ...string)` and let the caller pass everything in the correct order? Or did I misunderstand it?

File: lib/services/local/generic/generic_test.go (hunk 11 above)
Comment (codingllama) [reply lib/services/local/generic/generic_test.go:308]: They can still use s.MakeKey, right? I'm not sure I get the problem. If we can make MakeKey simpler I'm OK ditching the test, so this may all be a moot point.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto (hunk 14 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:138]: If we can change all member messages, then I think it's alright. If it's going to be a mix-and-match, then I'd rather keep it as-was (and save the comment for a future design ;) )

File: lib/services/local/access_list_test.go (hunk 9 above)
Comment (codingllama) [reply lib/services/local/access_list_test.go:110]: Thanks, Mike.  👍

File: lib/services/local/access_list.go (hunk 5 above)
Comment (codingllama) [reply lib/services/local/access_list.go]: Yes, agreed.

File: lib/services/local/access_list_test.go (hunk 9 above)
Comment (mdwn) [reply lib/services/local/access_list_test.go]: I split this into its own AccessListMembersCRUD function, let me know if that's looking alright.

File: lib/services/local/access_list.go (hunk 8 above)
Comment (mdwn) [reply lib/services/local/access_list.go]: I'm amenable to that, but I'd rather that be relegated to its own PR as it would cause a lot of changes for unrelated services in this one.

Comment (codingllama) [reply lib/services/local/access_list.go]: Yep, definitely not something to be changed here. Up to you if you think a follow up is worth it.

File: lib/services/local/generic/generic.go (hunk 10 above)
Comment (mdwn) [reply lib/services/local/generic/generic.go:293]: You didn't misunderstand it, but doing it that way creates an unfortunate amount of work on the caller. Basically, every caller would need to do this bit in order to get the proper ordering then:
```
	numKeyParts := len(parts)
	keyParts := make([]string, numKeyParts+2)
	keyParts[0] = s.backendPrefix
	keyParts[numKeyParts+1] = name
	if numKeyParts > 0 {
		copy(keyParts[1:], parts)
	}
        key = s.MakeKey(keyParts...)
```

I tried (a lot) to make it work the way you're describing and it ended up in a lot of duplication.

File: lib/services/local/generic/generic_test.go (hunk 11 above)
Comment (mdwn) [reply lib/services/local/generic/generic_test.go:308]: I don't think we can make `MakeKey` simpler, as per [this comment](https://github.com/gravitational/teleport/pull/30579#discussion_r1297511856). Let me know what you think.

**Review by flyinghermit — APPROVED** [review]

File: lib/services/local/generic/generic.go (hunk 10 above)
Comment (codingllama) [reply lib/services/local/generic/generic.go:293]: I still think it's doable, but looking back I think the naming/ordering threw me off.

Suggestion:

```go
// MakeKey returns a key in the format [s.backendPrefix, prefixes..., suffix].
func (s *Service[T]) MakeKey(suffix string, prefixes ...string) []byte {
	l := len(prefixes)
	parts := make([]string, l+2)
	parts[0] = s.backendPrefix
	if l > 0 {
		copy(parts[1:], prefixes)
	}
	parts[l+1] = name
	return backend.Key(parts...)
}
```

(and then ditch the MakeKey test?

File: lib/services/local/generic/generic.go [hunk 15]
```diff
@@ -97,8 +97,8 @@ func NewService[T types.Resource](cfg *ServiceConfig[T]) (*Service[T], error) {
 }
 
 // GetResources returns a list of all resources.
-func (s *Service[T]) GetResources(ctx context.Context) ([]T, error) {
-	rangeStart := backend.Key(s.backendPrefix, "")
+func (s *Service[T]) GetResources(ctx context.Context, prefixParts ...string) ([]T, error) {
```
Comment (codingllama) [inline lib/services/local/generic/generic.go:100]: MakeKey discussions aside, I'm finding this signature conceptually strange. Could you do a brief explanation of the key space we are dealing with and why this change is needed?

Comment (mdwn) [reply lib/services/local/generic/generic.go:100]: So most of (but not all) backend prefixes have "single" namespaces, e.g. the `access_lists` prefix is expected to contain all of the access list objects.

For access list members, we have the backend prefix `access_list_members`, but then these can be further namespaces by access list name. So `access_list_members/accessList1` or something like that. The idea here being that in addition to being able to refer to resources by just the backend prefix, you can now append arbitrary prefixes in case you want to further namespace these objects.

Does that make sense?

File: lib/services/local/generic/generic.go (hunk 10 above)
Comment (mdwn) [reply lib/services/local/generic/generic.go:293]: I can make the function change. What's the problem with the make key test, though? It's still nice to have this test coverage IMO.

Comment (codingllama) [reply lib/services/local/generic/generic.go]: No problem, it's just that MakeKey gets plenty of "exercise" from all other functions. If it's trivial to read, then I don't see much issue with it not getting a dedicated test.

File: lib/services/local/access_list_test.go [hunk 16]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+
+	members, _, err = service.ListAccessListMembers(ctx, accessList2.GetName(), 0, "")
+	require.NoError(t, err)
+	require.Empty(t, members)
+
+	// Verify access list members are not present.
+	accessList1Member1 := newAccessListMember(t, accessList1.GetName(), "1")
+	accessList1Member2 := newAccessListMember(t, accessList1.GetName(), "2")
+
+	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), "1")
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:170]: Maybe:

```suggestion
	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), accessList1Member1.GetName())
```

Ditto for others?

File: lib/services/local/access_list_test.go [hunk 17]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+	require.True(t, trace.IsNotFound(err))
+
+	// Add access list members.
+	_, err = service.UpsertAccessListMember(ctx, accessList1Member1)
+	require.NoError(t, err)
+	_, err = service.UpsertAccessListMember(ctx, accessList1Member2)
+	require.NoError(t, err)
+
+	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), "1")
+	require.NoError(t, err)
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:182]: Assert the 1st returned object?

File: lib/services/local/access_list_test.go [hunk 18]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), "2")
+	require.NoError(t, err)
+
+	accessList2Member1 := newAccessListMember(t, accessList2.GetName(), "2")
+	accessList2Member2 := newAccessListMember(t, accessList2.GetName(), "3")
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:187]: Did you go with "2" and "3" on purpose? Maybe use actual names instead of numbers?

File: lib/services/local/access_list_test.go [hunk 19]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+	accessList1Member2 := newAccessListMember(t, accessList1.GetName(), "2")
+
+	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), "1")
+	require.True(t, trace.IsNotFound(err))
+	_, err = service.GetAccessListMember(ctx, accessList1.GetName(), "2")
+	require.True(t, trace.IsNotFound(err))
+
+	// Add access list members.
+	_, err = service.UpsertAccessListMember(ctx, accessList1Member1)
+	require.NoError(t, err)
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:177]: Assert the 1st returned object?

File: lib/services/local/access_list_test.go [hunk 20]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+
+	accessList2Member1 := newAccessListMember(t, accessList2.GetName(), "2")
+	accessList2Member2 := newAccessListMember(t, accessList2.GetName(), "3")
+	_, err = service.UpsertAccessListMember(ctx, accessList2Member1)
+	require.NoError(t, err)
+	_, err = service.UpsertAccessListMember(ctx, accessList2Member2)
+	require.NoError(t, err)
+
+	// Fetch a paginated list of access lists members
+	paginatedMembers := make([]*accesslist.AccessListMember, 0, 2)
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:194]: Ditch the slightly-magical numbers?

```suggestion
	var paginatedMembers []*accesslist.AccessListMember
```

File: lib/services/local/access_list_test.go [hunk 21]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+	var nextToken string
+	for {
+		members, nextToken, err = service.ListAccessListMembers(ctx, accessList1.GetName(), 1, nextToken)
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:197]: Give "1" a comment or a name?

```suggestion
		const pageSize = 1
		members, nextToken, err = service.ListAccessListMembers(ctx, accessList1.GetName(), pageSize, nextToken)
```

File: lib/services/local/access_list_test.go [hunk 22]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
...
+
+		paginatedMembers = append(paginatedMembers, members...)
+		if nextToken == "" {
+			break
+		}
+	}
+	require.Empty(t, cmp.Diff([]*accesslist.AccessListMember{accessList1Member1, accessList1Member2}, paginatedMembers, cmpOpts...))
+
+	// Delete a member from an access list.
+	_, err = service.GetAccessListMember(ctx, accessList2.GetName(), "2")
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:208]:
```suggestion
	_, err = service.GetAccessListMember(ctx, accessList2.GetName(), accessList2Member1.GetName())
```

So it's less magical? Same for others.

File: lib/services/local/access_list_test.go [hunk 23]
```diff
@@ -125,6 +125,145 @@ func TestAccessListCRUD(t *testing.T) {
 	require.Empty(t, out)
 }
 
+func TestAccessListMembersCRUD(t *testing.T) {
```
Comment (codingllama) [inline lib/services/local/access_list_test.go:128]: Thanks, this is better.

I won't hold you on this, but I still think these do too much. In the future I'd consider splitting tests a bit more.

**Review by codingllama — COMMENTED** [review]
Looks good, just a few testing comments.

File: lib/services/local/access_list_test.go (hunk 18 above)
Comment (mdwn) [reply lib/services/local/access_list_test.go:187]: I did actually, to ensure that the member name didn't have to be unique. Will do, though.

File: lib/services/local/access_list_test.go (hunk 23 above)
Comment (mdwn) [reply lib/services/local/access_list_test.go:128]: Yeah, agreed. I tried to split it out for a bit but I realized it was going to take a pretty substantial amount of time, so I punted here.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 24]
```diff
@@ -124,15 +124,11 @@ message GetAccessListMemberRequest {
...
-  reserved "name";
+  reserved 1, 2, 3;
+  reserved "access_list", "name", "reason";
 
-  // access_list is the name of access list.
-  string access_list = 1;
-  // member_name is the name of the user to add.
-  string member_name = 4;
-  // reason is the reason the user is being added.
-  string reason = 3;
```
Comment (codingllama) [inline api/proto/teleport/accesslist/v1/accesslist_service.proto:105]: Is this field deletion going to cause problems to existing clients, or is this all new too?

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 25]
```diff
@@ -77,51 +80,69 @@ message GetAccessListRequest {
...
 // DeleteAccessListMemberRequest is the request for deleting a member from an access list.
 message DeleteAccessListMemberRequest {
+  // Renamed the "name" field to "member_name".
+  reserved 2;
+  reserved "name";
+
   // access_list is the name of access list.
   string access_list = 1;
-  // name is the name of the user to delete.
-  string name = 2;
```
Comment (codingllama) [inline api/proto/teleport/accesslist/v1/accesslist_service.proto:119]: Ditto for this change, any impact on existint clients?

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 26]
```diff
@@ -77,51 +80,69 @@ message GetAccessListRequest {
...
 // UpsertAccessListMemberRequest is the request for upserting an access list member.
 message UpsertAccessListMemberRequest {
-  // access_list is the name of access list.
-  string access_list = 1;
-  // name is the name of the user to add.
-  string name = 2;
-  // reason is the reason the user is being added.
-  string reason = 3;
-}
+  // Renamed the "name" field to "member_name".
```
Comment (codingllama) [inline api/proto/teleport/accesslist/v1/accesslist_service.proto:126]: nit: Not technically true, as name and member_name are distinct proto fields.

```suggestion
```

File: api/proto/teleport/accesslist/v1/accesslist_service.proto [hunk 27]
```diff
@@ -77,51 +80,69 @@ message GetAccessListRequest {
...
-message DeleteAccessListRequest {
-  // name is the name of the access list to delete.
-  string name = 1;
+  // member is the access list member to upsert.
+  Member member = 4;
 }
 
 // DeleteAccessListMemberRequest is the request for deleting a member from an access list.
 message DeleteAccessListMemberRequest {
+  // Renamed the "name" field to "member_name".
```
Comment (codingllama) [inline api/proto/teleport/accesslist/v1/accesslist_service.proto:136]:
```suggestion
```

**Review by codingllama — APPROVED** [review]
LGTM, assuming the proto deletions won't cause problems for existing clients.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto (hunk 24 above)
Comment (mdwn) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:105]: It's all new. It shouldn't cause any issues.

File: api/proto/teleport/accesslist/v1/accesslist_service.proto (hunk 25 above)
Comment (mdwn) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:119]: New as well.

**Review by codingllama — APPROVED** [review]

File: api/proto/teleport/accesslist/v1/accesslist_service.proto (hunk 24 above)
Comment (codingllama) [reply api/proto/teleport/accesslist/v1/accesslist_service.proto:105]: Thank you.

**Comment by mdwn** [comment]
> Thanks for the patience and sorry about the slow reviews, Mike.

No problem!
